[plugins] Add Charmed Kafka support

acsgn opened the pull request:
---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [ ] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [ ] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [ ] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?
- [X] Are all passwords or private data gathered by this PR [obfuscated](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin#how-to-prevent-collecting-passwords)?

phvalguima commented:
@acsgn Can you add @paulomach and @taurus-forever as reviewers to this PR? I cannot do so myself.

acsgn commented:
Hi @phvalguima I just resent the invitation to @paulomach. We just had a talk with @taurus-forever, he was on vacation last week, I hope he will respond hopefully by tomorrow. I pushed two new commits and trying to test the plugin. In addition, I am looking for a way to implement date filter for logs. I saw 3 different approaches and I am currently evaluating them. There is a --since option already, the add_journal method supports an until option and we also can use plugin options to manually achieve same functionality. I plan to use files' modified time attribute for filtering.

paulomach commented:
> Hi @phvalguima I just resent the invitation to @paulomach. We just had a talk with @taurus-forever, he was on vacation last week, I hope he will respond by tomorrow. I pushed two new commits and trying to test the plugin. In addition, I am looking for a way to implement date filter for logs. I saw 3 different approaches and I am currently evaluating them. There is a --since option already, the add_journal method supports an until option and we also can use plugin options to manually achieve same functionality. I plan to use files' modified time attribute for filtering.

paulomach commented:
@acsgn I've accidentally closed this, sorry!

I was meaning to comment that you need to invite @neppel to review it.

acsgn commented:
> @acsgn I've accidentally closed this, sorry!
> 
> I was meaning to comment that you need to invite @marceloneppel to review it.

No problem, sent an invite to @marceloneppel just now

marceloneppel reviewed (changes requested):
Hi @acsgn! You implemented the gathering of a lot of useful information. Thanks for that!

acsgn commented:
@marceloneppel Yes, you are right, I just changed Patroni to Patronictl and added the output from history. Also, do we run sosreport as sudo or not? It requires file permissions to read config file. So, I am not sure how to give permission to the tool.

phvalguima commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+            suggest_filename="metrics-patroni",
+        )
+
+        self.add_cmd_output(
```
Metrics are captured by Prometheus. We should pick up any config files for the exporter, but the metrics are already stored somewhere else.

phvalguima commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        # --- METRICS ---
+
+        self.add_cmd_output(
+            f"curl -vk {self.patroni_address}/metrics",
```
Metrics are captured by Prometheus. We should pick up any config files for the exporter, but the metrics are already stored somewhere else.

phvalguima commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        # followed by anything which may be enclosed with "
+        self.do_path_regex_sub(
+            f"{PATHS['PATRONI_CONF']}/*",
+            r'(password: )"?.*"?',
```
@marceloneppel is this the right format? This is the obfuscation part

marceloneppel replied:
Hi @phvalguima! It's right. I tested it at https://regex101.com/r/HbjFbO/1.

phvalguima commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+            f"{PATHS['PGBACKREST_LOGS']}",
+            f"{PATHS['PGBOUNCER_CONF']}",
+            f"{PATHS['PGBOUNCER_LOGS']}",
+        ])
```
Is there any specific files for the exporter service as well? If yes, can we capture them?

phvalguima commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        )
+
+    def postproc(self):
+        # --- SCRUB PASSWORDS ---
```
There is no passwords for the backup and exporter services as well?

acsgn replied:
I will check how the backup service authenticates, I am not collecting the exporter secrets but I can if requested,
`sudo snap get charmed-postgresql exporter`

acsgn replied:
pgbackrest.conf file is collected and cleaned accordingly

arif-ali commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        return "127.0.0.1:9187"
+
+    @property
+    def patroni_address(self) -> str | None:
```
This doesn't work with python3.6, and we need to ensure we have backward compatibility for that

arif-ali commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        return patroni_config["restapi"]["connect_address"]
+
+    @property
+    def patroni_cluster_name(self) -> str | None:
```
as above

arif-ali commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        return f"--config-file {PATHS['PATRONI_CONF']}/patroni.yml"
+
+    @property
+    def postgresql_host(self) -> str | None:
```
as above

arif-ali commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        return address.split(":")[0]
+
+    @property
+    def postgresql_port(self) -> str | None:
```
as above

arif-ali commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        return address.split(":")[1]
+
+    @property
+    def postgresql_username(self) -> str | None:
```
as above

arif-ali commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+        return superuser["username"]
+
+    @property
+    def postgresql_password(self) -> str | None:
```
as above

arif-ali commented on sos/report/plugins/charmed-postgresql.py:
```diff
@@ -0,0 +1,208 @@
+class CharmedPostgreSQL(Plugin, UbuntuPlugin):
+
+    short_desc = "Charmed PostgreSQL"
+    plugin_name = "charmed_postgresql"
```
May need to add the `packages` tuple to get this plugin triggered, this also works for snap packages too. The plugin may not be triggered without this potentially

```python
packages = ('charmed-postgresql',)
```

arif-ali reviewed (changes requested):
some initial comments

taurus-forever commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+
+    @property
+    def patroni_cluster_name(self) -> str:
+        with open(f"{PATHS['PATRONI_CONF']}/patroni.yml") as file:
```
It just doesn't work. STR:
```
juju deploy postgresql --channel 14/edge
juju ssh postgresql/0 bash


sudo wget -O /usr/lib/python3/dist-packages/sos/report/plugins/charmed_postgresql.py https://raw.githubusercontent.com/acsgn/sos/92ff2ded600939a45397b8ddeaca783e3cb21bbd/sos/report/plugins/charmed_postgresql.py


mkdir testsos && chmot 0777 testsos
sudo sos report     --only-plugins systemd,charmed_postgresql   --batch     --clean     --tmp-dir=./testsos     -z gzip

...
caught exception in plugin method "charmed_postgresql.setup()"
...


root@juju-8e5d32-1:/home/ubuntu/sosreport-juju-8e5d32-1-2024-04-05-xsgkali# cat sos_logs/charmed_postgresql-plugin-errors.txt
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sos/report/__init__.py", line 1224, in setup
    plug.setup()
  File "/usr/lib/python3/dist-packages/sos/report/plugins/charmed_postgresql.py", line 121, in setup
    f"topology {self.patroni_cluster_name}"),
  File "/usr/lib/python3/dist-packages/sos/report/plugins/charmed_postgresql.py", line 36, in patroni_cluster_name
    with open(f"{PATHS['PATRONI_CONF']}/patroni.yml") as file:
FileNotFoundError: [Errno 2] No such file or directory: '/var/snap/charmed-postgresql/current/etc/patroni/patroni.yml'

```

phvalguima replied:
@acsgn can you test these commands? I believe you will eventually hit an issue that refers to sosreport's policies if you do not run it within a snap. You can eventually circumvent that issue with:

```
diff --git a/sos/policies/distros/ubuntu.py b/sos/policies/distros/ubuntu.py
index f87ace69..a86188d7 100644
--- a/sos/policies/distros/ubuntu.py
+++ b/sos/policies/distros/ubuntu.py
@@ -42,9 +42,12 @@ class UbuntuPolicy(DebianPolicy):
             chroot=self.sysroot,
             remote_exec=remote_exec)
 
-        if self.package_manager.pkg_by_name(
-                'sosreport')['pkg_manager'] == 'snap':
-            self.sos_bin_path = '/snap/bin'
+        try:
+            if self.package_manager.pkg_by_name(
+                    'sosreport')['pkg_manager'] == 'snap':
+                self.sos_bin_path = '/snap/bin'
+        except TypeError:
+            self.sos_bin_path = ''
 
         self.valid_subclasses += [UbuntuPlugin]
```

acsgn replied:
Filename is updated and commands run as snap_daemon for read permission

taurus-forever reviewed (changes requested):
Are there some steps-to-test available to try this in our real charm?

```
juju ssh postgresql/0 bash

```

taurus-forever commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+        self.add_cmd_output(
+            (f"charmed-postgresql.patronictl {self.patronictl_args} "
+             f"topology {self.patroni_cluster_name}"),
+            suggest_filename="patroni-topology",
```
The generated content is useless :

```
root@juju-8e5d32-1:/home/ubuntu/sosreport-juju-8e5d32-1-2024-04-05-cytgbsp# cat sos_commands/charmed_postgresql/patroni-topology 
Error: Provided config file /var/snap/charmed-postgresql/current/etc/patroni/patroni.yaml not existing or no read rights. Check the -c/--config-file parameter
```

Please check the content for each file you collected in sosreport on the real charm. Tnx!

phvalguima replied:
@acsgn can you make sure each of the commands are functional against the snap?

taurus-forever commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+
+        # --- SNAP LOGS ---
+
+        self.add_journal("snap.charmed-postgresql.*")
```
The generated filename is fully:
```
-rw-r--r-- 1 root root 10011 Apr  5 14:03 journalctl_--no-pager_--unit_snap.charmed-postgresql
```

Is is expected?

arif-ali replied:
yes, this is expected

phvalguima commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+
+        self.add_journal("snap.charmed-postgresql.*")
+
+        # --- SNAP INFO ---
```
@taurus-forever that is already picked up by sosreport's snap plugin: https://github.com/sosreport/sos/blob/main/sos/report/plugins/snap.py#L32

arif-ali replied:
`snap services` is already collected by the `snap.py` plugin, so you would be grabbing duplicate information

taurus-forever reviewed (commented):
Q: should we remove IPs from configs?  In most of cases they are local... but not always (MaaS):

```
postgresql:
  listen: '99.62.99.58:5432'
  connect_address: '99.62.99.58:5432'
```

taurus-forever commented:
BTW, the password is leaking inside sosreport on data collection (IF user added `--verbose` flag?):

```
2024-04-05 14:03:46,183 DEBUG: [plugin:charmed_postgresql] packed command: binary=False, changes=False, chroot=True, cmd="PGPASSWORD=VtWblsblsbWfdfwefpOe charmed-postgresql.psql -U operator -h 99.62.99.58 -p 5432 -d postgres -c '\\dtS+'", cmd_as_tag=F
alse, container_cmd=None, env=None, foreground=False, priority=10, root_symlink=None, runat=None, sizelimit=25, stderr=True, subdir=None, suggest_filename='postgresql-tables', tags=[], timeout=None, to_file=False
```

+ OUPS in  ./var/snap/charmed-postgresql/common/var/log/postgresql/postgresql-5_1348.log
```
2024-04-05 13:48:42 UTC [10714]: user=operator,db=postgres,app=[unknown],client=99.62.99.58,line=1 LOG:  statement: CREATE ROLE "postgres"WITH LOGIN SUPERUSER ENCRYPTED PASSWORD 'heVctX222dummy32SKvLTk';
2024-04-05 13:48:42 UTC [10744]: user=operator,db=postgres,app=[unknown],client=99.62.99.58,line=1 LOG:  statement: CREATE ROLE "backup"WITH LOGIN SUPERUSER ENCRYPTED PASSWORD 'MRiT223dummy2322nHESx';
...
2024-04-05 13:48:57 UTC [12452]: user=operator,db=postgres,app=[unknown],client=99.62.99.58,line=1 LOG:  statement: CREATE ROLE "relation-4"WITH LOGIN ENCRYPTED PASSWORD 'gCTD123dummyasdanv4Ox' createrole createdb;
```

marceloneppel commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+            f"{PATHS['POSTGRESQL_LOGS']}",
+            f"{PATHS['PATRONI_CONF']}",
+            f"{PATHS['PATRONI_LOGS']}",
+            f"{PATHS['PGBACKREST_CONF']}",
```
I remembered that AWS S3 (or equivalent services) access and secret keys are stored in this file. It's needed to obfuscate them.

acsgn replied:
S3 secrets are obfuscated

taurus-forever commented:
JFYI, branched passwords in logs issue into charm issue: https://warthogs.atlassian.net/browse/DPE-4032
```
... LOG:  statement: CREATE ROLE "postgres" WITH LOGIN SUPERUSER ENCRYPTED PASSWORD 'heVctX222dummy32SKvLTk';
```

phvalguima commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+        with open(f"{PATHS['PATRONI_CONF']}/patroni.yml") as file:
+            patroni_config = yaml.safe_load(file)
+
+        address = patroni_config["postgresql"]["connect_address"]
```
```suggestion
    @property
    def postgresql_host(self) -> Optional[str]:
        with open(f"{PATHS['PATRONI_CONF']}/patroni.yml") as file:
            patroni_config = yaml.safe_load(file)

        address = patroni_config.get("postgresql", {}).get("connect_address", None)

        if not address:
            return None
        return address.split(":")[0]
```

acsgn replied:
Added Optional typing to properties

phvalguima commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+
+import yaml
+
+from sos.report.plugins import Plugin, UbuntuPlugin
```
Add `from typing import Optional`

acsgn commented on sos/report/plugins/charmed_postgresql.py:
```diff
@@ -0,0 +1,184 @@
+    "PGBACKREST_LOGS": SNAP_COMMON_PATH + "/var/log/pgbackrest",
+    "PGBOUNCER_CONF": SNAP_CURRENT_PATH + "/etc/pgbouncer",
+    "PGBOUNCER_LOGS": SNAP_COMMON_PATH + "/var/log/pgbouncer",
+}
```
Prometheus exporter logs are collected through journalctl
Pgbouncer support is added

acsgn commented:
How to test:

``` bash
multipass launch --cpus 4 --memory 8G --disk 20G --name testsos
mkdir ~/testsos
multipass mount ~/testsos testsos:~/testsos

multipass shell testsos
lxd init --auto
lxc network set lxdbr0 ipv6.address none
sudo snap install juju
mkdir -p ~/.local/share
juju bootstrap localhost overlord
juju add-model test

echo "default-base: ubuntu@22.04/stable
applications:
  data-integrator:
    charm: data-integrator
    channel: stable
    revision: 19
    num_units: 1
    to:
    - "0"
    options:
      database-name: test123
    constraints: arch=amd64
  postgresql:
    charm: postgresql
    channel: 14/stable
    revision: 363
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
    config:
      profile: testing
    storage:
      pgdata: rootfs,1,1024M
    trust: true
  s3-integrator:
    charm: s3-integrator
    channel: stable
    revision: 18
    num_units: 1
    to:
    - "0"
    options:
      bucket: s3demo
      endpoint: https://s3.eu-central-1.amazonaws.com
      path: /postgresql-test
      region: eu-central-1
    constraints: arch=amd64
  self-signed-certificates:
    charm: self-signed-certificates
    channel: stable
    revision: 72
    num_units: 1
    to:
    - "0"
    options:
      ca-common-name: Test CA
    constraints: arch=amd64
machines:
  "0":
    constraints: arch=amd64
relations:
- - postgresql:certificates
  - self-signed-certificates:certificates
- - postgresql:s3-parameters
  - s3-integrator:s3-credentials" > bundle.yaml

# Cannot include in the bundle due to base error
juju deploy pgbouncer --channel 1/stable
juju integrate postgresql pgbouncer
juju integrate data-integrator pgbouncer

juju run s3-integrator/leader sync-s3-credentials access-key=KEY secret-key=SECRET

juju ssh 0

sudo wget -O /usr/lib/python3/dist-packages/sos/report/plugins/charmed_postgresql.py https://raw.githubusercontent.com/acsgn/sos/feat/charmed-postgresql/sos/report/plugins/charmed_postgresql.py

mkdir -m 777 testsos
sudo sos report --only-plugins systemd,charmed_postgresql --batch --clean --tmp-dir=./testsos -z gzip
sudo chmod 666 testsos/*

exit

juju scp -- -r 0:~/testsos .

exit

multipass stop testsos
multipass delete --purge testsos
```

acsgn commented:
Example output of the plugin from the test steps can be seen at https://drive.google.com/drive/folders/1t1QOwzqwFTv140xvpaRpB5-kqRPOYth5?usp=sharing

marceloneppel reviewed (approved):
Thanks, @acsgn! LGTM!

taurus-forever reviewed (approved):
This is way better now and exactly what we need.
Well done! Tnx!
